---
title: 'Intro to phylogenetics: practical'
subtitle: 'AIMS-Imperial: modern statistics for global health'
author: "Alexandra Blenkinsop"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Objectives

This practical will show you how to run a phylogenetic pipeline from consensus sequences including creating an alignment, specifying a suitable outgroup for rooting, inferring a phylogeny using the distance-based neighbour-joining method and using a maximum-likelihood approach, resolving alignment issues and using bootstrapping to quantify uncertainty.

We will be using the *ape* (Analysis of Phylogenetics and Evolution) and *phangorn* packages for this practical, which have several tools for manipulating DNA sequence data in R and carrying out phylogenetic inference. The first step is to make sure you have these installed, as well as the other packages we need, using *install.packages()*.

# Load packages
```{r, include=TRUE, eval=TRUE, message=FALSE, echo=TRUE, warning=FALSE, tidy=TRUE}
require(data.table)
require(knitr)
# phylo packages
require(ape)
require(ade4)
require(phangorn)
require(phytools)
# packages for plotting
require(ggplot2)
require(ggtree)
require(ggsci)
require(viridis)

```

## Data

Download the consensus sequences for this practical from [github.com/MLGlobalHealth/aims_rwanda_2024](www.github.com/MLGlobalHealth/aims_rwanda_2024), which are a subset of partial polymerase (*pol*) sequences from HIV positive individuals from the Democratic Republic of Congo from this study by [Faria et al, 2019](https://doi.org/10.1371/journal.ppat.1007976), tracing the evolutionary dynamics of major HIV-1 subtypes across Central and Eastern Africa. The data were obtained from the [Los Alamos HIV sequence database](https://www.hiv.lanl.gov/components/sequence/HIV/search/search.html), by searching for the GenBank Accession numbers listed in the manuscript and selecting a subset of subtypes for the purpose of this practical.

To view the alignemt, you can download an alignment viewer such as  [AliView](https://ormbunkar.se/aliview/) (requires download) or  [jalview](https://www.jalview.org/jalview-js/JalviewJS.shtml/) (web interface). Here is a snapshot of part of the alignment:

```{r, include=TRUE, fig.height=8, fig.width=10, eval=TRUE, message=FALSE, echo=TRUE, warning=FALSE, tidy=TRUE}

# TODO: update these directories accordingly
data.dir <- '~/Documents/GitHub/aims_rwanda_2024/day1/practical3_phylo'
out.dir <- '~/Documents/GitHub/aims_rwanda_2024/day1/practical3_phylo'

knitr::include_graphics(file.path(data.dir,'faria_2019_alignment.png'))

```


The taxa names are named uniformly in the Los Alamos HIV sequence database, with the format <Subtype.Isocode.Year.Name.Accession>.

## Read in the alignment

Your first task is to load the multiple sequence alignment and have a look at its structure.

Read in the .fasta file. Note the format of FASTA - chevron denotes the taxon name. Gaps '-' are indel events (insertions/deletions). 'N' represents unknown characters.

How many sequences do we have and how long are they?


```{r, out.width='100%',include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# read data
filename <- file.path(data.dir,'faria_2019.fasta')

dat <- read.dna(file = filename, format = "fasta")
dat

```

Check the format of the data - the *ape* package in R handles objects with class DNAbin.

```{r, out.width='100%',include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

class(dat)

```

look at positions 1:10 in the alignment of the first 5 taxa

```{r, out.width='100%',include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}
as.character(dat)[1:5, 1:10]

```

We can also visualise these data in R, re-coding ambiguous sites as 'others'. The x-axis is the position in the alignment, but this is not meaningful unless it is mapped to some known genomic coordinates. The y-axis are the taxa. We show the first 30 taxa only.

```{r, out.width='100%',fig.height = 5, fig.width=10, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

par(mar=c(4,5,4,2))
ape::image.DNAbin(dat[1:30,], cex.lab = 0.4, cex.axis = 0.7)
```


## Generate the genetic distance matrix

Generate the distance matrix using the function dist.dna() from the *ape* package, assuming the Tamura and Nei model for the evolutionary rate, which has two parameters allowing for different base frequencies and different transition and substitution rates. This function generates a matrix of pairwise distances between taxa. Plot the matrix.

Note that entries $i=j$ have a genetic distance of zero.


```{r, out.width='100%', include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# calculate the pairwise genetic distances using Tamura and Nei model
D <- dist.dna(dat, model = "TN93")

# plot the resulting distance matrix
temp <- as.data.frame(as.matrix(D))
temp <- t(as.matrix(D))
temp <- temp[, ncol(temp):1]
par(mar = c(1, 5, 5, 1))

# melt the matrix to a data.table 
tmp <- data.table(reshape2::melt(temp))
# ensure the ordering of taxa are the same for plotting the matrix
tmp[, Var2:= factor(Var2,levels=levels(Var1))]

ggplot(tmp) + geom_tile(aes(x=Var1,y=Var2,fill=value)) +
  scale_fill_viridis(option="magma",direction=-1) + 
  labs(x='',y='',fill='Genetic distances\n(Tamura and Nei model)') +
  theme_bw(base_size=6) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
 
```

There are two types of HIV, HIV-1 and HIV-2, which can further be classified by subtype. All the sequences in this study are from individuals with HIV-1, which is the most prevalent type of HIV globally. The taxa labels indicate subtype. Subtypes are an indication of virus similarity. Which sequences have the smallest distances?

We find that sequences of the same subtype typically have smaller genetic distances than those of different subtypes.

## Build the tree

We would now like to visualise the relationship between these sequences in a more informative way by building a phylogeny.

We can use the *ape* package in R, which has a few algorithms available. Build a tree using the neighbour-joining algorithm.

```{r, out.width='100%', out.height='100%', fig.width=8, fig.height=10, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}
 
# build tree with neighbour-joining method
tre <- nj(D)
# trees which are generated using ape have class 'phylo'
class(tre)

summary(tre)

```

Plot the tree. What do you notice? Which tips share a common ancestor? Do any of the tree features suggest something has gone wrong with the alignment or phylogenetic inference?

```{r, out.width='100%', out.height='100%', fig.width=8, fig.height=10, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# map the taxa labels to colours based on their HIV subtype (for plotting)
select <- data.table(taxa=labels(dat))
select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

colmap <- data.table(ST=c('G','02_AG','A','Ref'),
                     color=mypal)
select <- merge(select,colmap,by='ST')
cols <- with(select, as.character(color[match(labels(dat), taxa)]))

# plot tree
plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Tree using neighbour-joining algorithm")

```


## Correct the alignment

See if you can find the reason for the long branch by taking a look at the alignment.

Open the .fasta file with an alignment viewer such as [AliView](https://ormbunkar.se/aliview/) [jalview](https://www.jalview.org/jalview-js/JalviewJS.shtml/) and see if you can find and correct the alignment error. Hint: the sequences in the alignment must all be the same length.

Export the amended alignment and save it with a different name in .fasta format.

Read in the new .fasta file and repeat the steps above using the following code. Now the distance matrix no longer shows large differences between all sequences and the misaligned sequence. The long branch has also gone from the tree.

What else do you notice? Hint: the tips of the tree have been coloured according to their HIV-1 subtype. 

 
```{r, out.width='100%', out.height='100%', include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}
filename <- file.path(data.dir,'faria_2019_fixedalignment.fasta')

dat <- read.dna(file = filename, format = "fasta")

D <- dist.dna(dat, model = "TN93")

# plot distance matrix
temp <- as.data.frame(as.matrix(D))
temp <- t(as.matrix(D))
temp <- temp[, ncol(temp):1]
par(mar = c(1, 5, 5, 1))

# melt the matrix to a data.table 
tmp <- data.table(reshape2::melt(temp))
# ensure the ordering of taxa are the same for plotting the matrix
tmp[, Var2:= factor(Var2,levels=levels(Var1))]

ggplot(tmp) + geom_tile(aes(x=Var1,y=Var2,fill=value)) +
  scale_fill_viridis(option="magma",direction=-1) + 
  labs(x='',y='',fill='Genetic distances\n(Tamura and Nei model)') +
  theme_bw(base_size=6) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r, out.width='100%', out.height='100%', fig.height=10, fig.width=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

select <- data.table(taxa=labels(dat))
select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

colmap <- data.table(ST=c('G','02_AG','A','Ref'),
                     color=mypal)
select <- merge(select,colmap,by='ST')
cols <- with(select, as.character(color[match(labels(dat), taxa)]))

# build tree
tre <- nj(D)

plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Tree using neighbour-joining algorithm")


```


We notice that there is a recombinant of subtypes A and G (light blue tip). The algorithm doesn't know where to put it in the tree, and we have a clade of both subtype A and subtype G viruses, whereas we'd expect the subtype A viruses to share a most recent common ancestor with all other subtype A sequences, and same for subtype G. What happens if we remove it?

Hint: the taxa labels for the alignment are stored in labels(dat).

Note that it is not necessary to exclude recombinants from phylogenetic analyses as a rule. However we have to be careful how we handle them. If we are interested in a particular recombinant circulating in the population, we might generate a separate phylogeny for just that recombinant.

# Remove the recombinant
```{r, out.width='100%',fig.width=8, fig.height=10, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# find recombinant taxa label which isn't subtype G or A
to_remove <- labels(dat)[!gsub('([A-Za-z0-9_]+).*','\\1',labels(dat)) %in% c('G','A','Ref')]

# remove recombinant
dat <- dat[!labels(dat)==to_remove,]

# tip colours
cols <- with(select, as.character(color[match(labels(dat), taxa)]))

# recompute distance matrix
D <- dist.dna(dat, model = "TN93")

# generate tree
tre <- nj(D)

# plot tree
plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Tree using neighbour-joining algorithm")


```


## Root the tree

The reference sequence (subtype C) is part of the G clade, and we still have some subtype A sequences in a clade with subtype G sequences. In general, the topology can be misleading if the tree is not rooted, especially with respect to isolates. We can tell the algorithm where the root should be by specifying the outgroup (the sequence beginning with *Ref.*).

We can root the tree using the command root().


```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# find the tip label containing 'Ref'
lab.root <- labels(dat)[grep('Ref',labels(dat))]

# re-root the tree
tre2 <- root(tre, outgroup = lab.root, resolve.root=TRUE)
plot(tre2, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Rooted tree using neighbour-joining algorithm")

```

Now by rooting the tree, the reference sequence has moved to be the ancestor of all the taxa. We now also see that all the subtype G taxa are in one clade, and all the subtype A taxa are in another.


## Extract a clade

There are a few different tools in *ape* for plotting. See if you can extract the G clade and plot it. You will need the following functions: fastMRCA() and extract.clade().

```{r, out.width='100%',fig.width=8, fig.height=4, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

node <- fastMRCA(tre2,"G.CD.2008.DRC319.MN178973",
    "G.CD.2008.DRC358.MN178994")
G.clade <- extract.clade(tre2,node)
plot(G.clade, cex = 0.6, tip.color = mypal[1])
add.scale.bar(x=0,y=0)
title("Subtype G clade")

```


## Bootstrapping

We are interested in seeing how robust our tree is.

Let's bootstrap resample each site of the alignment with replacement, carry out phylogenetic inference as before, and plot the tree with internal nodes labelled with the number of bootstrapped trees in which we observe this internal node. If the tree is robust, we would hope to observe the same internal nodes across the original tree and the bootstrap trees.

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# bootstrap 100 times
myBoots <- boot.phylo(tre2, dat, function(e) nj(dist.dna(e, model = "TN93")),B=100)

plot(tre2, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("NJ tree + bootstrap values")
nodelabels(myBoots, cex = 0.5)


```


What do you notice about the values?

Some of the bootstrap values at the internal nodes are quite low, which means some of these nodes are not well supported.

## Re-build tree using maximum likelihood estimation

We will finally re-estimate the tree using a maximum-likelihood approach ([Felsenstein, 1981](https://link.springer.com/article/10.1007/BF01734359)), which uses the sequence data in a statistical framework where we seek to estimate model parameters. There other approaches to building trees we will not cover here, such as maximum parsimony, but you can explore these yourself.

We will use the *phangorn* package, which requires the data to be in a different format, phyDat.

We first compute the likelihood of the data using the neighbour-joining tree using the command pml().


```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# convert the alignment to PhyDat using phangorn package
dat2 <- as.phyDat(dat)
dat2
tre2.lik <- pml(tre2, dat2)
tre2.lik

```

We next optimise, where the algorithm searches the tree topology space to find the best tree, optimising tree topology (optNni), base frequencies (optBf) and substitution rates (optQ).

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

opt.tre <- optim.pml(tre2.lik, optNni = TRUE, optBf = TRUE, optQ = TRUE,
 optGamma = TRUE)
opt.tre

```

We could also optimise for a given substitution model, with the option: model = 'JC' or 'GTR'.

We can then compare the optimal tree with the initial tree with a likelihood ratio test and confirm that it has a substantially larger likelihood and a lower AIC, indicating the optimised tree is better model for the sequence data.

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

anova(tre2.lik, opt.tre)

```


We can plot it again,

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

opt.tre.root <- root(opt.tre$tree, outgroup = lab.root, resolve.root=TRUE)
plot(opt.tre.root, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Rooted maximum likelihood tree")


```

We can also re-compute the bootstrap values for internal nodes.

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center", warning=FALSE}


bs <- bootstrap.pml(opt.tre, bs=100, optNni=TRUE, multicore=TRUE, control = pml.control(trace=0))
plotBS(midpoint(opt.tre$tree), bs, type="p",cex = 0.6, tip.color = cols)
add.scale.bar()
title("ML tree")

```


